hazelcast/src/main/java/com/hazelcast/query/impl/UnorderedIndexStore.java

```diff
@@ -97,6 +143,7 @@ public void clear() {
                 if (value == NULL) {
                     records = recordsWithNullValue;
                 } else {
+                    // value is already canonicalized by the associated index
```

Comment:
Minor: should we canonicalize arguments for other methods as well in the calling method? 

For instance, instead of canonicalizing in `UnorderedIndexStore#getRecords(java.lang.Comparable, boolean, java.lang.Comparable, boolean)`, canonicalize in `AbstractIndex#getRecords(java.lang.Comparable, boolean, java.lang.Comparable, boolean)` instead? Same for `UnorderedIndexStore#getRecords(java.lang.Comparable)` and `AbstractIndex#getRecords(java.lang.Comparable)`. 

Does it depend on the index type?

Reply:
I was thinking about that too and yes, it depends on the index type, or more precisely, on the index store type. It's cleaner to move all the canonicalization logic to `AbstractIndex`, but that requires a lot of refactoring, otherwise there will be either a lot of code duplicated between `AbstractIndex` and `BaseIndexStore` or we still will have some cross-dependencies.

For instance, `BaseIndexStore.unwrapAndInsertToIndex`/`unwrapAndRemoveFromIndex`/`sanitizeValue` must be moved to `AbstractIndex`, but we have to track the index stats for `MultiResult`, but `IndexStore.insert`/`remove` returns `void`, so we can't track the stats on the `AbstractIndex` level, so we have to adjust the return type to know the removed entries at least, but we can't avoid passing `MultiResult` to `insert`/`remove`, because `multiResultHasToDetectDuplicates` tracking is based on `MultiResult`, but we would have two rounds of iteration in this case, one in `AbstractIndex` and one in `BaseIndexStore` which would look mostly identical. So in the end I decided to keep all the logic in the index stores and make indexes to behave just like gateways to the underlying store implementations.

I think `IndexStore` implementations must be cleaned up in a separate task, all the logic unrelated to the actual storing must be moved to the `Index` level, but what would be left in the stores after the cleanup then? Probably there is no much sense in their existence after all, or probably we should do it in reverse: move all the logic to stores :)